src/v/cloud_storage/recovery_utils.cc

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/recovery_utils.h"
+
+#include "cloud_storage/logger.h"
+
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace cloud_storage {
+
+ss::sstring
+generate_result_path(const storage::ntp_config& ntp_cfg, bool result) {
+    auto uuid = boost::uuids::random_generator()();
+    return fmt::format(
+      recovery_result_format,
+      recovery_result_prefix,
+      ntp_cfg.ntp().ns(),
+      ntp_cfg.ntp().tp.topic(),
+      ntp_cfg.ntp().tp.partition,
+      boost::uuids::to_string(uuid),
+      result);
+}
+
+ss::future<> place_download_result(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  const storage::ntp_config& ntp_cfg,
+  bool result_completed,
+  retry_chain_node& parent) {
+    retry_chain_node fib{&parent};
+    auto result_path = generate_result_path(ntp_cfg, result_completed);
+    auto result = co_await remote.upload_object(
+      bucket, cloud_storage_clients::object_key{result_path}, "", fib);
+    if (result != upload_result::success) {
+        vlog(
+          cst_log.error,
+          "failed to upload result to {}, error: {}",
+          result_path,
+          result);
```

Comment:
is it expected that the recovery will proceed if it can't put the lock file to S3? maybe it'll be better to throw and stop at this point?

Reply:
At this point in the code, we have completed creating the topics and are waiting for downloads. The file uploaded is a result file. It is used by the download watcher to tally if the required downloads are complete. Once all result files are available, it will finish the recovery and reset the topic retention parameters. 

But a failure to upload the result here is hard to recover from, because we have already recovered the data for the partition replica on the node. The failure effectively here means the recovery process will never be able to finish, it will keep waiting for the result from this replica. The end results will be:

1. no other recovery can be run, becuase the current recovery keeps waiting for the replica download result which we were unable to upload.
2. the topic retention is stuck on the temporary retention used by the recovery process - this can be fixed by manually resetting the retention using rpk etc.

For fixing issue `1` there is no easy way at this point (short of placing a result file manually in the bucket, which will stop the recovery process the next time the counts are checked). Maybe it can be implemented as part of the cancellation operation for recovery which was discussed recently.